### ISSUE TITLE
Error routing messages (loss & no delivry for all reachable destinations while - and after that - one of the destinations is unreachable)

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#1664]
Error routing messages  with Mirth 2.0.1 channels (loss & no delivry for all reachable destinations while - and after that - one of the destinations is unreachable)
Use case: (6 Mirth channels: A,B,C,D,E,F: with properties: store message data:yes, state:enabled, synchronize channel:yes)

channel A: (router)  (source: hl7 2.5 jdbc Oracle) routing messages to channels B,C,D,E,F  for all of them: destination type: channel Writer, Wait for channel response: no)
channel B: destination type: channel writer (channel name=none)    (here only control channel)
channels C & D: destination type: MLLP with a host & port reachable.
channels E & F: destination type: MLLP with a host (temporarily) non reachable (as it is the case after a network failure)

For channels C,D,E,F: keep connection open: no, maximum retry count:6, use persistent queues:no, ACK Timeout: 15s Process HL7 ACK: Yes Send Response to: None
Channel A was feeded (for test) by a 50 messages to read in table,  polled every 5 secondes. (the "Run on update statement" = No  (only for this test to simulate the load).
Channel A was manualy stopped after 3550 messages received, with only 1073 queued (the error) and only 1073 distributed to each of the the reachable channels: B,C,D.
For the unreachabled channels: E,F: after a moment the Errored stated reached 1073 and Queued statistics on channel A reached to 0.
=> we should obain 3550 messages received on channels B,C & D instead of 1073.
After a reset of the message table, clear statistics, deleting the mule queues and restart the server, a second test was conducted with the same result.

The same test was conducted with not problem on another Mirth 1.8.2 plateform. (postgres 8.4/ Windows)
More over, with Mirth 2.0.1, in such situation, it becomes very difficult, sometime not possible to stop channel A without disable the channel & restart Mirth server.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-1675
Reporter: bruno.martin@chu-dijon.fr
Created: 2010-11-25T10:04:41.000-0800

**Comment by rbeckman-nextgen** [comment]
chanels used for the test

Imported Comment. Original Details:
Author: bruno.martin@chu-dijon.fr
Created: 2010-11-25T10:44:01.000-0800

**Comment by rbeckman-nextgen** [comment]
In fact: 2 use cases: First( one destination is not reachable with grants pb on it),  Second (the host is not reachable: simple network failure) . A bug different in the two cases. 1rst : lost of messages, 2nd: all destinations too slow (waiting retries on unreachable destination)

Imported Comment. Original Details:
Author: bruno.martin@chu-dijon.fr
Created: 2010-11-25T11:32:24.000-0800

**Comment by rbeckman-nextgen** [comment]
Does this happen if you keep the queue under 1000?  The max queue size in 2.0 defaults to 1000, so generally it will start blocking when it hits 1001.  This has been resolved in MIRTH-1694 and we will release a patch to adjust this back to 100,000 like it was in 1.8.2.

Imported Comment. Original Details:
Author: jacobb
Created: 2010-12-16T17:28:40.000-0800

**Comment by rbeckman-nextgen** [comment]
We did not encountered this problem with queues less than 1000 items.

However, with Mirth 1.8.2 as 2.0.1, when the maximum element of the mule queue is reached (100000 elements with Mirth 1.8.2, 1000 with Mirth 2.0.1), a 1000+1 th element (100000+1 th element) is actually accepted by entering the channel but will never be distributed, even if the size of the queue decreases thereafter due to the consumption of messages by the output channel (when network access is restored to the MLLP channel output).
After decay of the queue, we find that the messages that follow the 1000+1 are correctly routed to the output channel (as the queue is not saturated) while the 1000+1 th message remains undistributed (idem with a 100000+1 th item with Mirth 1.8.2).


Best regard,

Bruno Martin
Chef de Projet Infrastructure EAI
Direction des Systèmes d'Information
Centre Hospitalier Universitaire de Dijon
Tél : 03.80.29.31.90
Mail : bruno.martin@chu-dijon.fr
Site web CHU : www.chu-dijon.fr



-----Message d'origine-----
De : Jacob Brauer (JIRA) [mailto:issues@mirthcorp.com]
Envoyé : vendredi 17 décembre 2010 02:30
À : MARTIN Bruno
Objet : [JIRA] Commented: (MIRTH-1675) Error routing messages (loss & no delivry for all reachable destinations while - and after that - one of the destinations is unreachable)


    [ http://www.mirthcorp.com/community/issues/browse/MIRTH-1675?page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel&focusedCommentId=14311#action_14311 ]

Jacob Brauer commented on MIRTH-1675:
-------------------------------------

Does this happen if you keep the queue under 1000?  The max queue size in 2.0 defaults to 1000, so generally it will start blocking when it hits 1001.  This has been resolved in MIRTH-1694 and we will release a patch to adjust this back to 100,000 like it was in 1.8.2.


--
This message is automatically generated by JIRA.
-
If you think it was sent incorrectly contact one of the administrators: http://www.mirthcorp.com/community/issues/secure/Administrators.jspa
-
For more information on JIRA, see: http://www.atlassian.com/software/jira




Imported Comment. Original Details:
Author: bruno.martin@chu-dijon.fr
Created: 2010-12-17T00:35:06.000-0800

**Comment by rbeckman-nextgen** [comment]
This issue should have been resolved by MIRTH-1677 and MIRTH-1694 with the default queue size now set to 0, or infinite.  The problem seems to only occur after the queue size has reached its limit.  After it has reached its limit, it will start blocking.  The channel has a receiving in memory buffer that will continue to hold messages it is receiving and waiting to process.  Once that fills up it will begin rejecting messages (never receiving them or allowed the socket connections to be established), but that generally takes hundreds of messages.  

I tried to reproduce this issue with both the channel writer queue and the LLP sender queue, and did not experience message loss with either connector.  For both connectors I hit the limit, it queued one past it, and then the received count went up by one more as it started to buffer all messages after that.  For example, with a queue limit of 5, received = 7 and queued = 6.  I sent 9 messages, and the following two were buffered and waiting to be received.  As soon as I started up the destination, the counts on both channels were correct at 9 received and 9 sent.  Issues like the one described in this bug report might happen as more messages are sent, but the new queue limit at 0 should resolve the problem entirely.

Imported Comment. Original Details:
Author: jacobb
Created: 2010-12-17T13:49:49.000-0800
